src/mem/alloc.rs

```diff
@@ -0,0 +1,1063 @@
+            .encode_order(self.order)
+            .encode_next(self.next_page)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let next = ((mem.0 & PAGE_FREE_NEXT_MASK) >> PAGE_FREE_NEXT_SHIFT) as usize;
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        FreeInfo {
+            next_page: next,
+            order: order,
+        }
+    }
+}
+
+struct AllocatedInfo {
+    order: usize,
+}
+
+impl AllocatedInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_ALLOCATED).encode_order(self.order)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        AllocatedInfo { order: order }
+    }
+}
+
+struct SlabPageInfo {
+    slab: VirtAddr,
+}
+
+impl SlabPageInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::encode_slab(self.slab)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        SlabPageInfo {
+            slab: VirtAddr::new(mem.0 & PAGE_TYPE_SLABPAGE_MASK),
+        }
+    }
+}
+
+struct CompoundInfo {
+    order: usize,
+}
+
+impl CompoundInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_COMPOUND).encode_order(self.order)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        CompoundInfo { order: order }
+    }
+}
+
+struct ReservedInfo {}
+
+impl ReservedInfo {
+    fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_RESERVED)
+    }
+
+    pub fn decode(_mem: PageStorageType) -> Self {
+        ReservedInfo {}
+    }
+}
+
+enum Page {
+    Free(FreeInfo),
+    Allocated(AllocatedInfo),
+    SlabPage(SlabPageInfo),
+    CompoundPage(CompoundInfo),
+    Reserved(ReservedInfo),
+}
+
+impl Page {
+    pub fn to_mem(&self) -> PageStorageType {
+        match self {
+            Page::Free(fi) => fi.encode(),
+            Page::Allocated(ai) => ai.encode(),
+            Page::SlabPage(si) => si.encode(),
+            Page::CompoundPage(ci) => ci.encode(),
+            Page::Reserved(ri) => ri.encode(),
+        }
+    }
+
+    pub fn from_mem(mem: PageStorageType) -> Self {
+        let page_type = mem.0 & PAGE_TYPE_MASK;
+
+        if page_type == PAGE_TYPE_FREE {
+            Page::Free(FreeInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_ALLOCATED {
+            Page::Allocated(AllocatedInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_SLABPAGE {
+            Page::SlabPage(SlabPageInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_COMPOUND {
+            Page::CompoundPage(CompoundInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_RESERVED {
+            Page::Reserved(ReservedInfo::decode(mem))
+        } else {
+            panic!("Unknown Page Type {}", page_type);
+        }
+    }
+}
+
+struct MemoryRegion {
```

Comment:
Please, add a comment explaining we use MemoryRegions to provide Buddy algorithm